codec/unknownproto/unknown_fields.go

```diff
@@ -0,0 +1,377 @@
+package unknownproto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+type Checker struct {
+	// AllowUnknownNonCriticals when set will skip over non-critical fields that are unknown.
+	AllowUnknownNonCriticals bool
+}
+
+// By default, the defaultChecker will report every single unknown field, whether non-critical or not.
+// To customize this behavior, create a custom checker and modify its behavior.
+var defaultChecker = &Checker{AllowUnknownNonCriticals: false}
+
+// CheckMismatchedFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the unknown fields by tagNumber and wireType, or mismatched wireTypes.
+// If you'd like to toggle between allowing non-critical unknown fields, use CheckMismatchedFieldsAllowNonCritical
+func CheckMismatchedFields(b []byte, msg proto.Message) error {
+	return defaultChecker.CheckMismatchedFields(b, msg)
+}
+
+func (ckr *Checker) CheckMismatchedFields(b []byte, msg proto.Message) error {
```

Comment:
```suggestion
func (ckr *Checker) RejectUnknownFields(b []byte, msg proto.Message) error {
```

As I've said before, checking for mismatched fields is a non-goal.

Reply:
Done, thanks.